File: 프로그래머스/다트 게임.py
def solution(dartResult):
    lis = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]
    answer = 0
    # 문자열에서 현재 위치
    idx = 0
    # 바로 직전 점수
    beforeScore = 0
    # 이번 점수가 두자리 수인지 판별
    twoCheck = 0
    # 두자리 수인지 체크해서 인덱스에 반영
    while idx < len(dartResult):
        if dartResult[idx+1] in lis:
            twoCheck = 1
        else:
            twoCheck = 0
        
        score = int(dartResult[idx : idx + 1 + twoCheck])
        bonus = dartResult[idx + 1 + twoCheck]
        # 옵션 존재하는 경우
        if (idx + 2 + twoCheck) != len(dartResult) and (dartResult[idx + 2 + twoCheck] == "*" or dartResult[idx + 2 + twoCheck] == "#"):
            option = dartResult[idx + 2 + twoCheck]
            # 인덱스 옮겨줌
            idx += 3 + twoCheck
            # 스타상인 경우
            if option == "*":
                if bonus == "S":
                    answer += (score * 2)
                    answer += beforeScore
                    beforeScore = (score * 2)
                elif bonus == "D":
                    answer += ((score ** 2) * 2)
                    answer += beforeScore
                    beforeScore = ((score ** 2) * 2)
                else:
                    answer += ((score ** 3) * 2)
                    answer += beforeScore
                    beforeScore = ((score ** 3) * 2)
            # 아차상인 경우
            else:
                if bonus == "S":
                    answer -= score 
                    beforeScore = -score
                elif bonus == "D":
                    answer -= (score ** 2)
                    beforeScore = -(score ** 2)
                else:
                    answer -= (score ** 3)
                    beforeScore = -(score ** 3)
        # 옵션 없는 경우              
        else:
            # 인덱스 옮겨줌
            idx += 2 + twoCheck
            if bonus == "S":
                answer += score
                beforeScore = score
            elif bonus == "D":
                answer += (score ** 2)
                beforeScore = (score ** 2)
            else:
                answer += (score ** 3)
                beforeScore = (score ** 3)
    return answer

File: 프로그래머스/다트 게임.java
public class Solution {
    public int solution(String dartResult) {
        int answer = 0, before = 0, now = 0;
        for (int i = 0; i < dartResult.length(); i++) {
            char ch = dartResult.charAt(i);
            // 점수
            if (Character.isDigit(ch)) {
                // 점수가 두자리수 = 10점 경우
                if (ch == '1' && dartResult.charAt(i+1) == '0') {
                    now = 10;
                    i++;
                } else
                    now = ch - '0';
                // 보너스
            } else if (Character.isUpperCase(ch)) {
                if (ch == 'D')
                    now = (int) Math.pow(now, 2);
                else if (ch == 'T')
                    now = (int) Math.pow(now, 3);
                // 옵션
            } else {
                if (ch == '*') {
                    answer += before;
                    now *= 2;
                } else
                    now *= -1;
            }
            // 마지막 점수 구간인 경우 or 다음 글자가 숫자인 경우 하나의 점수 구간이 완료되므로 총합 계산, 직전 점수 갱신
            if (i == dartResult.length()-1 || Character.isDigit(dartResult.charAt(i+1))) {
                answer += now;
                before = now;
            }
        }
        return answer;
    }
}